isthmus/src/main/java/io/substrait/isthmus/SubstraitRelNodeConverter.java

```diff
@@ -89,22 +109,128 @@ public RelNode visit(NamedScan namedScan) throws RuntimeException {
+          case INNER -> JoinRelType.INNER;
+          case LEFT -> JoinRelType.LEFT;
+          case RIGHT -> JoinRelType.RIGHT;
+          case OUTER -> JoinRelType.FULL;
+          case SEMI -> JoinRelType.SEMI;
+          case ANTI -> JoinRelType.ANTI;
+          case UNKNOWN -> throw new UnsupportedOperationException(
+              "Unknown join type is not supported");
+        };
+    return relBuilder.push(left).push(right).join(joinType, condition).build();
+  }
+
+  @Override
+  public RelNode visit(Aggregate aggregate) throws RuntimeException {
+    RelNode child = aggregate.getInput().accept(this);
+    var groupExprLists =
+        aggregate.getGroupings().stream()
             .map(
-                expr ->
-                    expr.accept(
-                        new ExpressionRexConverter(
-                            relOptCluster.getTypeFactory(), scalarFunctionConverter)))
+                gr ->
+                    gr.getExpressions().stream()
+                        .map(expr -> expr.accept(expressionRexConverter))
+                        .toList())
             .toList();
+    List<RexNode> groupExprs =
+        groupExprLists.stream().flatMap(Collection::stream).collect(Collectors.toList());
+    RelBuilder.GroupKey groupKey = relBuilder.groupKey(groupExprs, groupExprLists);
 
-    return relBuilder.push(child).project(rexList).build();
+    List<AggregateCall> aggregateCalls =
+        aggregate.getMeasures().stream().map(m -> fromMeasure(m)).toList();
+    return relBuilder.push(child).aggregate(groupKey, aggregateCalls).build();
+  }
+
+  private AggregateCall fromMeasure(Aggregate.Measure measure) {
+    var arguments =
+        measure.getFunction().arguments().stream()
+            .map(expr -> expr.accept(expressionRexConverter))
+            .toList();
+    var operator =
+        aggregateFunctionConverter.getSqlOperatorFromSubstraitFunc(
+            measure.getFunction().declaration().key(), measure.getFunction().outputType());
+    if (!operator.isPresent()) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unable to find binding for call %s", measure.getFunction().declaration().name()));
+    }
+    List<Integer> argIndex = new ArrayList<>();
+    for (RexNode arg : arguments) {
+      if (!(arg instanceof RexInputRef)) {
+        throw new UnsupportedOperationException(
+            String.format(
+                "aggregation function only support column as argument. % is not support", arg));
+      }
+      argIndex.add(((RexInputRef) arg).getIndex());
+    }
+
+    boolean distinct =
+        measure.getFunction().invocation()
+            == AggregateFunction.AggregationInvocation.AGGREGATION_INVOCATION_DISTINCT;
+
+    SqlAggFunction aggFunction;
+    RelDataType returnType =
+        TypeConverter.convert(relOptCluster.getTypeFactory(), measure.getFunction().getType());
+
+    if (operator.get() == SqlStdOperatorTable.COUNT) {
```

Comment:
This looks like it should be turned into something. I also don't really understand it. It looks like you're saying if a == foo, set b to foo. Why not just set b to foo in the cases a is one of the defined items.

Reply:
I changed to:
```
   if (operator.get() instanceof SqlAggFunction) {
      aggFunction = (SqlAggFunction) operator.get();
    }
```

`operator` is `SqlOperator`,  a base class.  I could not directly assign it to `aggFunction`.